open_spiel/bots/stockfish/stockfish_bot.cc

```diff
@@ -0,0 +1,137 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "open_spiel/bots/stockfish/stockfish/src/bitboard.h"
+#include "open_spiel/bots/stockfish/stockfish/src/endgame.h"
+#include "open_spiel/bots/stockfish/stockfish/src/position.h"
+#include "open_spiel/bots/stockfish/stockfish/src/search.h"
+#include "open_spiel/bots/stockfish/stockfish/src/thread.h"
+#include "open_spiel/bots/stockfish/stockfish/src/tt.h"
+#include "open_spiel/bots/stockfish/stockfish/src/uci.h"
+#include "open_spiel/bots/stockfish/stockfish/src/syzygy/tbprobe.h"
+#include "open_spiel/bots/stockfish/stockfish/src/types.h"
+#include "open_spiel/bots/stockfish/stockfish_bot.h"
+#include "open_spiel/spiel_utils.h"
+
+#include "open_spiel/games/chess.h"
+
+namespace open_spiel {
+namespace stockfish {
+namespace {
+
+// FEN string of the initial position, normal chess
+const char* StartFEN = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1";
+
+
+}  // namespace
+
+StockfishBot::StockfishBot(int argc, char **argv, int move_time) :
+  move_time_(move_time) {
+
+
+  CommandLine::init(argc, argv);
+  UCI::init(Options);
+  Tune::init();
+  PSQT::init();
+  Bitboards::init();
+  Position::init();
+  Bitbases::init();
+  Endgames::init();
+  Threads.set(size_t(Options["Threads"]));
+  Search::clear(); // After threads are up
+  Eval::init_NNUE();
+
+  if (Options.count("Use NNUE")) {
+    Options["Use NNUE"] = std::string("false");
+  }
+}
+
+void StockfishBot::Restart() {
+  Search::clear(); // ucinewgame command
+}
+
+void StockfishBot::RestartAt(const State& state) {
+  Restart();
+}
+
+
+Action StockfishBot::Step(const State& state) {
+
+  auto chess_state = dynamic_cast<const chess::ChessState *>(&state);
+
+  Position pos;
+  auto sf_states_ = StateListPtr(new std::deque<StateInfo>(1)); // Drop old and create a new one
+  pos.set(state.ToString(), false, &sf_states_->back(), Threads.main());
+
+
+  Search::LimitsType limits;
+  limits.startTime = now();
+  limits.movetime = move_time_;
+
+  Threads.start_thinking(pos, sf_states_, limits, false);
+
+  Threads.main()->wait_for_search_finished();
+
+  Thread* bestThread = Threads.get_best_thread();
+
+  // Send again PV info if we have a new best thread
+  if (bestThread != Threads.main())
+    sync_cout << UCI::pv(bestThread->rootPos, bestThread->completedDepth, -VALUE_INFINITE, VALUE_INFINITE) << sync_endl;
+
+  sync_cout << "bestmove " << UCI::move(bestThread->rootMoves[0].pv[0], false);
+
+  std::cout << sync_endl;
+
+  Move best_move = bestThread->rootMoves[0].pv[0];
+
+  chess::Move move = MoveFromStockfishMove(best_move, chess_state->Board());
+
+  return chess::MoveToAction(move);
+}
+
+chess::Move StockfishBot::MoveFromStockfishMove(Move move, const chess::StandardChessBoard &board) {
```

Comment:
We will need some better namespace control... it's hard to distinguish openspiel chess move from stockfish Move. I guess they didn't put their code into a namespace... Maybe you can put all of their header files into a namespace ::stockfish, something like this (I didn't try this):
```
namespace stockfish {
#include "open_spiel/bots/stockfish/stockfish/src/bitboard.h"
#include "open_spiel/bots/stockfish/stockfish/src/endgame.h"
#include "open_spiel/bots/stockfish/stockfish/src/position.h"
#include "open_spiel/bots/stockfish/stockfish/src/search.h"
#include "open_spiel/bots/stockfish/stockfish/src/thread.h"
#include "open_spiel/bots/stockfish/stockfish/src/tt.h"
#include "open_spiel/bots/stockfish/stockfish/src/uci.h"
#include "open_spiel/bots/stockfish/stockfish/src/syzygy/tbprobe.h"
#include "open_spiel/bots/stockfish/stockfish/src/types.h"
}
```
 and rename this namespace `open_spiel::stockfish` to `open_spiel::bots`. Then we will have nice chess::Move and stockfish::Move names, and stockfish:: will always refer to the original implementation.

Reply:
It's hard for me to review this code without the namespaces as I don't know what is in OS and in SF

Reply:
(It will likely not lead to anything, but just in case, I asked in https://github.com/official-stockfish/Stockfish/issues/3350 whether stockfish could be into a namespace.)